pkg/agent/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -646,6 +649,21 @@ func (c *Controller) syncRule(key string) error {
 		return nil
 	}
 
+	if !(c.antreaProxyEnabled && c.proxyAllEnabled) && rule.isAppliedToServicesIngressPolicyRule() {
+		klog.InfoS("Ingress policy rule applied to Services is valid only when both AntreaProxy and proxyAll are enabled, skipping", "ruleID", key)
+		if c.statusManagerEnabled {
+			c.statusManager.SetRuleAsInvalid(key, rule.PolicyUID, rule.PolicyName, rule.Name, "AntreaProxy or ProxyAll is not enabled")
+		}
+		return nil
+	}
+	if !c.antreaProxyEnabled && rule.isToServicesEgressPolicyRule() {
+		klog.InfoS("Egress policy rule to Services is valid only when AntreaProxy is enabled, skipping", "ruleID", key)
+		if c.statusManagerEnabled {
+			c.statusManager.SetRuleAsInvalid(key, rule.PolicyUID, rule.PolicyName, rule.Name, "AntreaProxy is not enabled")
+		}
+		return nil
+	}
```

Comment:
I think we shouldn't add these implementatic specific checks in controller. Such special handling could eventually make `syncRule` extremely long and very specific to implementation, especially when the controller is extended to support more scenarios, like Node policy.

It should be the reconciler instead of the controller to do such check as it's the one knowing how to implement the rule. The reconciler should return an error to controller when it's unable to implement it and the controller should just report the error in a generic way to antrea-controller, without the need to know what's the error. In this way, the workflow can be reused by any realization failure in the future.

Reply:
Just found I commented the same before: https://github.com/antrea-io/antrea/pull/5591#discussion_r1369892288

Reply:
Do you meant that we should do such checks in `reconciler.go`.  If a rule is invalid, it should return an error, then we report the error to antrea-controller through statusController? If so, do we still need to requeue the rule? If statusController is not enabled, what should we do? Just log something?